C/rsort.c

```diff
@@ -70,64 +64,202 @@ static void swap(const sha256_midstate** a, const sha256_midstate** b) {
+ *
+ * Precondition: For all begin <= i < end, NULL != a[i];
+ *               for all i < CHAR_COUNT, bucketSize[i] is the number of of entries in the subentries in a[begin:end) whose ix-th character is 'i'.
+ *               for all i < UMAX_CHAR, bucketEdge[i] + bucketSize[i] = bucketEdge[i+1];
+ *               ix < sizeof((*a)->s);
+ * Postcondition: The contents of bucketSize is not preserved!
+ */
+static void sort_buckets(const sha256_midstate** a, uint32_t* restrict bucketSize, const uint32_t* restrict bucketEdge, unsigned int ix) {
+  /* The implementation works by finding the first non-empty bucket and then swapping the first element of that bucket into its position
+     at the far end of the bucket where it belongs.
+
+     After moving the element into that position the size of the target bucket is decreased by one,
+     and thus the next element that will be swapped into that bucket will be placed behind it.
+
+     This process continues until the first element of the first non-empty bucket gets swapped with itself
+     and its own bucket is decremented from size 1 to size 0.
+
+     At that point we search again for the next non-empty bucket and repeat this process until there are no more non-empty buckets.
+   */
+  for (unsigned int i = 0; i < CHAR_COUNT; ++i) {
+    size_t start = bucketEdge[i];
+    while (bucketSize[i]) {
+      /* Each time through this while loop some bucketSize is decremented.
+         Therefore this body is executed 'sum(i < CHAR_COUNT, bucketSize[i]) = end - begin' many times.
+       */
+      size_t bucket = readIndex(a[start], ix);
+      simplicity_assert(bucketSize[bucket]);
+      bucketSize[bucket]--;
+      swap(a + start, a + bucketEdge[bucket] + bucketSize[bucket]);
+    }
+  }
+}
+
+/* Attempts to (partially) sort an array of pointers to 'sha256_midstate's in place in memcmp order.
+ * If NULL != hasDuplicates and if duplicate entries are found, the sorting is aborted and
+ * *hasDuplicates is set to one of pointers of a duplicate entry.
+ * If NULL != hasDuplicates and no duplicate entries are found, then *hasDuplicates is set to NULL.
  *
- * The maximum recursion depth is 'level'.  With some effort 'rsort' could be rewritten to be non-recursive.
  * The time complexity of rsort is O('len').
  *
- * Precondition: size_t scratch[(level + 1)*CHAR_COUNT];
- *               For all 0 <= i < len, NULL != a[i];
- *               level <= sizeof((*a)->s);
+ * Precondition: For all 0 <= i < len, NULL != a[i];
  */
-const sha256_midstate* rsort(size_t* scratch, const sha256_midstate** a, size_t len, size_t level) {
-  /* An array of length 0 or 1 is sorted and without duplicates. */
-  if (len < 2) return NULL;
+static void rsort_ex(const sha256_midstate** a, uint_fast32_t len, const sha256_midstate** hasDuplicates) {
+  uint32_t stack[(CHAR_COUNT - 1)*(sizeof(*a)->s) + 1]; /* approximately 32 KiB */
```

Comment:
There is an open question as to whether this should be allocated on the stack or on the heap.  Previously it was heap allocated.  In this PR it is stack allocated.

I ran this code through https://github.com/PeterMcKinnis/WorstCaseStack to get some perspective on this issue.
Note: I had to manually comment out jet evaluation.  Because jets are "dynamically dispatched" the stack analysis cannot know how much stack space jets will use.  So the numbers we have below are without any jets being used.  Also note: The analysis also doesn't know how much stack space `free`, `malloc` etc use, and the values also don't include whatever stack space is used by these functions.

The analysis of this PR suggests that `rsort` uses 34056 bytes of stack and `elements_simplicity_execSimplicity` uses 34784 bytes of stack.

If we were to heap allocate this, instead then `rsort` would use roughly  1432 bytes of stack and `elements_simplicity_execSimplicity` would use 2160 bytes of stack.

Relatedly, the worst case stack use of any of our current jets appears to be `check_sig_verify` which uses 5704 bytes of stack, with `secp256k1_ecmult` taking up 4632 bytes of stack

One thing to be mindful of is low-power hardware devices wanting to evaluate Simplicity.  For example Jade hardware had by default a very small stack, and they perform various stack tricks at the moment to support stack intensive operations required to do secp-zkp operations like verifying rangeproofs.

Reply:
@apoelstra Unless you object, I think I'll update this PR this to being heap allocated.

Reply:
Yeah, I think that's best. Staying under 8K stack will make a big difference to the portability of the code. And I don't think this library can work without malloc anyway, so there's no portability cost to heap allocation...just a code review/maintenance cost.

Reply:
It is now heap allocated.